{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bbi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': 249250621,\n",
       " 'chr10': 135534747,\n",
       " 'chr11': 135006516,\n",
       " 'chr12': 133851895,\n",
       " 'chr13': 115169878,\n",
       " 'chr14': 107349540,\n",
       " 'chr15': 102531392,\n",
       " 'chr16': 90354753,\n",
       " 'chr17': 81195210,\n",
       " 'chr18': 78077248,\n",
       " 'chr19': 59128983,\n",
       " 'chr2': 243199373,\n",
       " 'chr20': 63025520,\n",
       " 'chr21': 48129895,\n",
       " 'chr22': 51304566,\n",
       " 'chr3': 198022430,\n",
       " 'chr4': 191154276,\n",
       " 'chr5': 180915260,\n",
       " 'chr6': 171115067,\n",
       " 'chr7': 159138663,\n",
       " 'chr8': 146364022,\n",
       " 'chr9': 141213431,\n",
       " 'chrM': 16571,\n",
       " 'chrX': 155270560,\n",
       " 'chrY': 59373566}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bw_map = '/mnt/beegfs/userdata/j_wang/Algorithm_Resources/dig/Mappability/wgEncodeCrgMapabilityAlign36mer.bigWig'\n",
    "bbi.chromsizes(bw_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class TabixFile in module pysam.libctabix:\n",
      "\n",
      "class TabixFile(pysam.libchtslib.HTSFile)\n",
      " |  Random access to bgzf formatted files that\n",
      " |  have been indexed by :term:`tabix`.\n",
      " |  \n",
      " |  The file is automatically opened. The index file of file\n",
      " |  ``<filename>`` is expected to be called ``<filename>.tbi``\n",
      " |  by default (see parameter `index`).\n",
      " |  \n",
      " |  Parameters\n",
      " |  ----------\n",
      " |  \n",
      " |  filename : string\n",
      " |      Filename of bgzf file to be opened.\n",
      " |  \n",
      " |  index : string\n",
      " |      The filename of the index. If not set, the default is to\n",
      " |      assume that the index is called ``filename.tbi``\n",
      " |  \n",
      " |  mode : char\n",
      " |      The file opening mode. Currently, only ``r`` is permitted.\n",
      " |      \n",
      " |  parser : :class:`pysam.Parser`\n",
      " |  \n",
      " |      sets the default parser for this tabix file. If `parser`\n",
      " |      is None, the results are returned as an unparsed string.\n",
      " |      Otherwise, `parser` is assumed to be a functor that will return\n",
      " |      parsed data (see for example :class:`~pysam.asTuple` and\n",
      " |      :class:`~pysam.asGTF`).\n",
      " |  \n",
      " |  encoding : string\n",
      " |  \n",
      " |      The encoding passed to the parser\n",
      " |  \n",
      " |  threads: integer\n",
      " |      Number of threads to use for decompressing Tabix files.\n",
      " |      (Default=1)\n",
      " |  \n",
      " |  \n",
      " |  Raises\n",
      " |  ------\n",
      " |  \n",
      " |  ValueError\n",
      " |      if index file is missing.\n",
      " |  \n",
      " |  IOError\n",
      " |      if file could not be opened\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      TabixFile\n",
      " |      pysam.libchtslib.HTSFile\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __reduce__ = __reduce_cython__(...)\n",
      " |      TabixFile.__reduce_cython__(self)\n",
      " |  \n",
      " |  __setstate__ = __setstate_cython__(...)\n",
      " |      TabixFile.__setstate_cython__(self, __pyx_state)\n",
      " |  \n",
      " |  close(self)\n",
      " |      TabixFile.close(self)\n",
      " |      \n",
      " |      closes the :class:`pysam.TabixFile`.\n",
      " |  \n",
      " |  fetch(self, reference=None, start=None, end=None, region=None, parser=None, multiple_iterators=False)\n",
      " |      TabixFile.fetch(self, reference=None, start=None, end=None, region=None, parser=None, multiple_iterators=False)\n",
      " |      fetch one or more rows in a :term:`region` using 0-based\n",
      " |              indexing. The region is specified by :term:`reference`,\n",
      " |              *start* and *end*. Alternatively, a samtools :term:`region`\n",
      " |              string can be supplied.\n",
      " |      \n",
      " |              Without *reference* or *region* all entries will be fetched. \n",
      " |              \n",
      " |              If only *reference* is set, all reads matching on *reference*\n",
      " |              will be fetched.\n",
      " |      \n",
      " |              If *parser* is None, the default parser will be used for\n",
      " |              parsing.\n",
      " |              \n",
      " |              Set *multiple_iterators* to true if you will be using multiple\n",
      " |              iterators on the same file at the same time. The iterator\n",
      " |              returned will receive its own copy of a filehandle to the file\n",
      " |              effectively re-opening the file. Re-opening a file creates\n",
      " |              some overhead, so beware.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods defined here:\n",
      " |  \n",
      " |  __new__(*args, **kwargs) from builtins.type\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  contigs\n",
      " |      list of chromosome names\n",
      " |  \n",
      " |  filename_index\n",
      " |  \n",
      " |  header\n",
      " |      the file header.\n",
      " |      \n",
      " |      The file header consists of the lines at the beginning of a\n",
      " |      file that are prefixed by the comment character ``#``.\n",
      " |      \n",
      " |      .. note::\n",
      " |          The header is returned as an iterator presenting lines\n",
      " |          without the newline character.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __pyx_vtable__ = <capsule object NULL>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pysam.libchtslib.HTSFile:\n",
      " |  \n",
      " |  __enter__(self)\n",
      " |      HTSFile.__enter__(self)\n",
      " |  \n",
      " |  __exit__(self, exc_type, exc_value, traceback)\n",
      " |      HTSFile.__exit__(self, exc_type, exc_value, traceback)\n",
      " |  \n",
      " |  add_hts_options(self, format_options=None)\n",
      " |      HTSFile.add_hts_options(self, format_options=None)\n",
      " |      Given a list of key=value format option strings, add them to an open htsFile\n",
      " |  \n",
      " |  check_truncation(self, ignore_truncation=False)\n",
      " |      HTSFile.check_truncation(self, ignore_truncation=False)\n",
      " |      Check if file is truncated.\n",
      " |  \n",
      " |  get_reference_name(self, tid)\n",
      " |      HTSFile.get_reference_name(self, tid)\n",
      " |      \n",
      " |      return :term:`contig` name corresponding to numerical :term:`tid`\n",
      " |  \n",
      " |  get_tid(self, contig)\n",
      " |      HTSFile.get_tid(self, contig)\n",
      " |      \n",
      " |      return the numerical :term:`tid` corresponding to\n",
      " |      :term:`contig`\n",
      " |      \n",
      " |      returns -1 if contig is not known.\n",
      " |  \n",
      " |  is_valid_reference_name(self, contig)\n",
      " |      HTSFile.is_valid_reference_name(self, contig)\n",
      " |      \n",
      " |      return True if the contig name :term:`contig` is valid; False otherwise.\n",
      " |  \n",
      " |  is_valid_tid(self, tid)\n",
      " |      HTSFile.is_valid_tid(self, tid)\n",
      " |      \n",
      " |      return True if the numerical :term:`tid` is valid; False otherwise.\n",
      " |      \n",
      " |      returns -1 if contig is not known.\n",
      " |  \n",
      " |  parse_region(self, contig=None, start=None, stop=None, region=None, tid=None, reference=None, end=None)\n",
      " |      HTSFile.parse_region(self, contig=None, start=None, stop=None, region=None, tid=None, reference=None, end=None)\n",
      " |      parse alternative ways to specify a genomic region. A region can\n",
      " |              either be specified by :term:`contig`, `start` and\n",
      " |              `stop`. `start` and `stop` denote 0-based, half-open\n",
      " |              intervals. :term:`reference` and `end` are also accepted for\n",
      " |              backward compatibility as synonyms for :term:`contig` and\n",
      " |              `stop`, respectively.\n",
      " |      \n",
      " |              Alternatively, a samtools :term:`region` string can be\n",
      " |              supplied.\n",
      " |      \n",
      " |              If any of the coordinates are missing they will be replaced by\n",
      " |              the minimum (`start`) or maximum (`stop`) coordinate.\n",
      " |      \n",
      " |              Note that region strings are 1-based inclusive, while `start`\n",
      " |              and `stop` denote an interval in 0-based, half-open\n",
      " |              coordinates (like BED files and Python slices).\n",
      " |      \n",
      " |              If `contig` or `region` or are ``*``, unmapped reads at the end\n",
      " |              of a BAM file will be returned. Setting either to ``.`` will\n",
      " |              iterate from the beginning of the file.\n",
      " |      \n",
      " |              Returns\n",
      " |              -------\n",
      " |      \n",
      " |              tuple :\n",
      " |                  a tuple of `flag`, :term:`tid`, `start` and\n",
      " |                  `stop`. The flag indicates whether no coordinates were\n",
      " |                  supplied and the genomic region is the complete genomic space.\n",
      " |      \n",
      " |              Raises\n",
      " |              ------\n",
      " |      \n",
      " |              ValueError\n",
      " |                 for invalid or out of bounds regions.\n",
      " |  \n",
      " |  reset(self)\n",
      " |      HTSFile.reset(self)\n",
      " |      reset file position to beginning of file just after the header.\n",
      " |      \n",
      " |              Returns\n",
      " |              -------\n",
      " |      \n",
      " |              The file position after moving the file pointer. : pointer\n",
      " |  \n",
      " |  seek(self, offset, whence=0)\n",
      " |      HTSFile.seek(self, uint64_t offset, int whence=io.SEEK_SET)\n",
      " |      move file pointer to position *offset*, see :meth:`pysam.HTSFile.tell`.\n",
      " |  \n",
      " |  tell(self)\n",
      " |      HTSFile.tell(self)\n",
      " |      return current file position, see :meth:`pysam.HTSFile.seek`.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from pysam.libchtslib.HTSFile:\n",
      " |  \n",
      " |  category\n",
      " |      General file format category.  One of UNKNOWN, ALIGNMENTS,\n",
      " |      VARIANTS, INDEX, REGIONS\n",
      " |  \n",
      " |  closed\n",
      " |      return True if HTSFile is closed.\n",
      " |  \n",
      " |  compression\n",
      " |      File compression.\n",
      " |      \n",
      " |      One of NONE, GZIP, BGZF, CUSTOM.\n",
      " |  \n",
      " |  description\n",
      " |      Vaguely human readable description of the file format\n",
      " |  \n",
      " |  duplicate_filehandle\n",
      " |  \n",
      " |  filename\n",
      " |  \n",
      " |  format\n",
      " |      File format.\n",
      " |      \n",
      " |      One of UNKNOWN, BINARY_FORMAT, TEXT_FORMAT, SAM, BAM,\n",
      " |      BAI, CRAM, CRAI, VCF, BCF, CSI, GZI, TBI, BED.\n",
      " |  \n",
      " |  index_filename\n",
      " |  \n",
      " |  is_bam\n",
      " |      return True if HTSFile is reading or writing a BAM alignment file\n",
      " |  \n",
      " |  is_bcf\n",
      " |      return True if HTSFile is reading or writing a BCF variant file\n",
      " |  \n",
      " |  is_closed\n",
      " |      return True if HTSFile is closed.\n",
      " |  \n",
      " |  is_cram\n",
      " |      return True if HTSFile is reading or writing a BAM alignment file\n",
      " |  \n",
      " |  is_open\n",
      " |      return True if HTSFile is open and in a valid state.\n",
      " |  \n",
      " |  is_read\n",
      " |      return True if HTSFile is open for reading\n",
      " |  \n",
      " |  is_remote\n",
      " |  \n",
      " |  is_sam\n",
      " |      return True if HTSFile is reading or writing a SAM alignment file\n",
      " |  \n",
      " |  is_stream\n",
      " |  \n",
      " |  is_vcf\n",
      " |      return True if HTSFile is reading or writing a VCF variant file\n",
      " |  \n",
      " |  is_write\n",
      " |      return True if HTSFile is open for writing\n",
      " |  \n",
      " |  mode\n",
      " |  \n",
      " |  threads\n",
      " |  \n",
      " |  version\n",
      " |      Tuple of file format version numbers (major, minor)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pysam.TabixFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pysam.libctabix.TabixFile at 0x7f56ef64eb00>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pysam.TabixFile(\"/mnt/beegfs/userdata/j_wang/Algorithm_Resources/dig/downloads/mutation_files/PCAWG/ICGC_only/Biliary-AdenoCA_SNV_MNV_INDEL.ICGC.annot.bed.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<HDF5 dataset \"Biliary-AdenoCA\": shape (10000,), type \"<i8\">\n",
      "<HDF5 dataset \"idx\": shape (10000, 3), type \"<i8\">\n",
      "<HDF5 dataset \"x_data\": shape (10000, 10000, 2), type \"<f8\">\n"
     ]
    }
   ],
   "source": [
    "with h5py.File('/mnt/beegfs/scratch/j_wang/03_Results/Dig/h5/002_split_data_10000_0_0.8.h5') as fd:\n",
    "    for k in fd.keys():\n",
    "        print(fd[k])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RepDigDrive",
   "language": "python",
   "name": "repdigdrive"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
